tir/tokenProvider.go

```diff
@@ -0,0 +1,232 @@
+
+type TokenProvider interface {
+	GetToken(vc *verifiable.Credential, audience string) (string, error)
+	GetAuthCredential() (vc *verifiable.Credential, err error)
+}
+
+type M2MTokenProvider struct {
+	// encodes the token according to the configuration
+	tokenEncoder TokenEncoder
+	// the credential
+	authCredential *verifiable.Credential
+	// the signing key
+	signingKey *rsa.PrivateKey
+	// clock to get issuance time from
+	clock common.Clock
+	// verification method to be used on the tokens
+	verificationMethod string
+	// signature type to be used
+	signatureType string
+	// type of the provided key
+	keyType string
+	// did of the token provider
+	did string
+}
+
+type TokenEncoder interface {
+	GetEncodedToken(vp *verifiable.Presentation, audience string) (encodedToken string, err error)
+}
+
+type Base64TokenEncoder struct{}
+
+func InitM2MTokenProvider(config *configModel.Configuration, clock common.Clock) (tokenProvider TokenProvider, err error) {
+	m2mConfig := config.M2M
+
+	if m2mConfig.KeyPath == "" {
+		logging.Log().Warn("No private key configured, cannot provide m2m tokens.")
+		return tokenProvider, ErrorTokenProviderNoKey
+	}
+	if m2mConfig.VerificationMethod == "" {
+		logging.Log().Warn("No verification method configured, cannot provide m2m tokens.")
+		return tokenProvider, ErrorTokenProviderNoVerificationMethod
+	}
+
+	privateKey, err := getSigningKey(m2mConfig.KeyPath)
+	if err != nil {
+		logging.Log().Warnf("Was not able to load the signing key. Err: %v", err)
+		return tokenProvider, err
+	}
+
+	if m2mConfig.CredentialPath == "" {
+		return tokenProvider, ErrorTokenProviderNoVC
+	}
+	if config.Verifier.Did == "" {
+		logging.Log().Warn("No did for token provider")
+		return tokenProvider, ErrorTokenProviderNoDid
+	}
+
+	vc, err := getCredential(m2mConfig.CredentialPath)
+	if err != nil {
+		logging.Log().Warnf("Was not able to load the credential. Err: %v", err)
+		return tokenProvider, err
+	}
+	logging.Log().Debug("Successfully initialized the M2MTokenProvider.")
+	return M2MTokenProvider{tokenEncoder: Base64TokenEncoder{}, authCredential: vc, signingKey: privateKey, did: config.Verifier.Did, clock: clock, verificationMethod: m2mConfig.VerificationMethod, keyType: config.M2M.KeyType, signatureType: config.M2M.SignatureType}, err
+}
+
+func (tokenProvider M2MTokenProvider) GetAuthCredential() (vc *verifiable.Credential, err error) {
+	return tokenProvider.authCredential, err
+}
+
+func (tokenProvider M2MTokenProvider) GetToken(vc *verifiable.Credential, audience string) (token string, err error) {
+
+	vp, err := tokenProvider.signVerifiablePresentation(vc)
+	if err != nil {
+		logging.Log().Warnf("Was not able to get a signed verifiable presentation. Err: %v", err)
+		return token, err
+	}
+	return tokenProvider.tokenEncoder.GetEncodedToken(vp, audience)
+}
+
+func (base64TokenEncoder Base64TokenEncoder) GetEncodedToken(vc *verifiable.Presentation, audience string) (encodedToken string, err error) {
```

Comment:
Do we need the audience for anything here?

Reply:
Not in the base64 encoded token, but we will need it for the jwt-tokens, thus I added it to the interface